atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTopologyValidator.java

```diff
@@ -143,23 +147,86 @@ Set<CassandraServer> getNewHostsWithInconsistentTopologies(
+        // If a consensus can be reached from the current servers, filter all new servers which have the same set of
+        // host ids. Accept dissent as such, but permit
+        ClusterTopologyResult topologyFromCurrentServers =
+                maybeGetConsistentClusterTopology(currentServersWithoutSoftFailures);
+        switch (topologyFromCurrentServers.type()) {
+            case CONSENSUS:
+                Preconditions.checkState(
+                        topologyFromCurrentServers.agreedTopology().isPresent(),
+                        "Expected to have a consistent topology for a CONSENSUS result, but did not.");
+                ConsistentClusterTopology topology =
+                        topologyFromCurrentServers.agreedTopology().get();
+                pastConsistentTopology.set(topology);
+                return EntryStream.of(newServersWithoutSoftFailures)
                         .removeValues(result -> result.type() == HostIdResult.Type.SUCCESS
                                 && result.hostIds().equals(topology.hostIds()))
                         .keys()
-                        .toSet())
-                .orElseGet(newServersWithoutSoftFailures::keySet);
+                        .toSet();
+            case DISSENT:
+                // In the event of *active* dissent, we want to hard fail.
+                return newServersWithoutSoftFailures.keySet();
+            case NO_QUORUM:
+                // In the event of no quorum, we trust the new servers iff they agree with our historical knowledge
+                // of what the old servers were thinking, since in containerised deployments all nodes can change
+                // between refreshes for legitimate reasons (but they should still refer to the same underlying
+                // cluster).
+                if (pastConsistentTopology.get() == null) {
+                    // We don't have a record of what worked in the past, so just reject.
+                    return newServersWithoutSoftFailures.keySet();
+                }
+                Optional<ConsistentClusterTopology> maybeTopology = maybeGetConsistentClusterTopology(
+                                newServersWithoutSoftFailures)
+                        .agreedTopology();
+                if (maybeTopology.isEmpty()) {
+                    log.info(
+                            "No quorum was detected among old servers, and new servers were also not in"
+                                    + " agreement. Not adding new servers in this case.",
+                            SafeArg.of("newServers", CassandraLogHelper.collectionOfHosts(newlyAddedHosts)),
+                            SafeArg.of("allServers", CassandraLogHelper.collectionOfHosts(allHosts.keySet())));
+                    return newServersWithoutSoftFailures.keySet();
+                }
+                ConsistentClusterTopology newNodesAgreedTopology = maybeTopology.get();
+                if (!newNodesAgreedTopology
```

Comment:
Yep - I assume your concern here is that two sets that have the same elements but different order might end up not-being-equal, but `equals()` is actually [defined](https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#equals(java.lang.Object)) to avoid this kind of issue:
> Compares the specified object with this set for equality. Returns true if the specified object is also a set, the two sets have the same size, and every member of the specified set is contained in this set (or equivalently, every member of this set is contained in the specified set). This definition ensures that the equals method works properly across different implementations of the set interface.

Reply:
Nice, TIL!